patches/master_patch.patch

```diff
@@ -1831,6 +1831,488 @@ index a433afd0217830b49ddb3381ad55afeae2381840..ddcca749666f956c6317676dbcdd48da
+ 
+   bool is_valid() const { return scheme_ != SCHEME_INVALID; }
+ 
+@@ -67,7 +69,12 @@ class NET_EXPORT ProxyServer {
+   // Returns true if this ProxyServer is a QUIC proxy.
+   bool is_quic() const { return scheme_ == SCHEME_QUIC; }
+ 
++  // Returns true if this ProxyServer has username/password authentication
++  bool is_auth() const { return auth_; }
++
+   const HostPortPair& host_port_pair() const;
++  const std::string& username() const;
++  const std::string& password() const;
+ 
+   // Parses from an input with format:
+   //   [<scheme>"://"]<server>[":"<port>]
+@@ -146,15 +153,18 @@ class NET_EXPORT ProxyServer {
+ 
+   bool operator==(const ProxyServer& other) const {
+     return scheme_ == other.scheme_ &&
+-           host_port_pair_.Equals(other.host_port_pair_);
++           host_port_pair_.Equals(other.host_port_pair_) &&
++           username_ == other.username_ &&
++           password_ == other.password_;
+   }
+ 
+   bool operator!=(const ProxyServer& other) const { return !(*this == other); }
+ 
+   // Comparator function so this can be placed in a std::map.
+   bool operator<(const ProxyServer& other) const {
+-    return std::tie(scheme_, host_port_pair_) <
+-           std::tie(other.scheme_, other.host_port_pair_);
++    return std::tie(scheme_, host_port_pair_, auth_, username_, password_) <
++           std::tie(other.scheme_, other.host_port_pair_,
++                    other.auth_, other.username_, other.password_);
+   }
+ 
+   // Returns the estimate of dynamically allocated memory in bytes.
+@@ -168,8 +178,15 @@ class NET_EXPORT ProxyServer {
+       std::string::const_iterator host_and_port_begin,
+       std::string::const_iterator host_and_port_end);
+ 
++  // Returns the `<user>:<pass>@' prefix, if authentication is
++  // enabled, or empty if not.
++  std::string AuthString() const;
++
+   Scheme scheme_;
+   HostPortPair host_port_pair_;
++  bool auth_;
++  std::string username_;
++  std::string password_;
+ };
+ 
+ typedef std::pair<HostPortPair, ProxyServer> HostPortProxyPair;
+diff --git a/net/socket/client_socket_pool_manager.cc b/net/socket/client_socket_pool_manager.cc
+index ffec16ff2944915301b2af4ca34b556cbc177b5d..07dcb8111b833292cf59148514064a1356b2b70b 100644
+--- a/net/socket/client_socket_pool_manager.cc
++++ b/net/socket/client_socket_pool_manager.cc
+@@ -13,6 +13,7 @@
+ #include "net/proxy/proxy_info.h"
+ #include "net/socket/client_socket_handle.h"
+ #include "net/socket/client_socket_pool.h"
++#include "net/socket/socks_auth_username_password.h"
+ #include "net/socket/socks_client_socket_pool.h"
+ #include "net/socket/ssl_client_socket_pool.h"
+ #include "net/socket/transport_client_socket_pool.h"
+@@ -189,9 +190,20 @@ int InitSocketPoolHelper(ClientSocketPoolManager::SocketGroupType group_type,
+       connection_group = base::StringPrintf(
+           "socks%c/%s", socks_version, connection_group.c_str());
+ 
+-      socks_params = new SOCKSSocketParams(proxy_tcp_params,
```

Comment:
I think this patch can be minimized by not bothering with `is_auth()` and just always passing `proxy_server.username()` and `proxy_server.password()`
don't assign to temp vars because that just increases the patch size